assets/js/modules/forms.js

```diff
@@ -30,6 +31,28 @@ function handleAbandonmentMessage(formEl) {
     });
 }
 
+/*
+ * Warns the user before leaving a page if the form's state is different from when the page was loaded.
+ *
+ * Caveats: form must have the class below and the field must lose focus to trigger the change() event.
+ * */
+function warnOnUnsavedChanges() {
+    let formHasChanged = false;
+    $('.js-form-warn-unsaved').each(function() {
+        const $form = $(this);
+        const initialState = $form.serialize();
+        $(':input', $form).change(function() {
```

Comment:
Curious if rather than adding a change listener to every form element is it possible to serialise the form _within_ `beforeunload` method and do the check there, or does the return value for `beforeunload` always get used for the prompt?

Reply:
Tried this – was kind of possible (`beforeunload` can return `false` which stops the message appearing), but meant we had to always bind `beforeunload` (eg. as soon as the page loads) which Chrome complains about, and meant immediately overriding it (eg. for form submits etc) so felt a bit circular.

Reply:
I did wonder when I asked that if it was even possible, beforeunload is a thorny one, thanks for checking though.